### PR TITLE
Moving Camera update and render frame to Game Loop (Main thread)

### PR DESCRIPTION
**Pull request by samcake**
This PR modify the function and thread place where the Camera and MyAvatar are updated every frame.
They are now updated in the main thread  in the Application::update() call in the Game loop. And not in the render thread just before running the render engine frame.

To get there we changed the following components:
- Introduced a struct "AppRenderArgs" to pass the values needed by render loop from game loop to configure the engine. The struct is updated by the game loop in a thread safe section (in Application::update()) and execute the code that has been removed from the Render Loop: Camera Update and last minute MyAvatar update.

- In order to guaranty a "Game Frame" boundary we introduced a  frame granularity on the transactions pushed to the scene. at the end of the Game loop, we call main3DScene->enqueueFrame() which mark the end of this game frame and flush the transactions in the queue onward to another "framed" queue ready for render thread. The render loop processes all the transactions in the "framed" queue to update render items with the state captured during the game frame of the transactions.

- Moved the functions of Application supposed to be called from the render thread in a separate file "Application_render.cpp" in an effort to make Application easier to navigate maybe....

And cleaned up a few things:
- removing  legacy calls on Application.
- fixing the Debug Anim Draw  render item which was still updated after being deleted.

## TEST PLAN
The overall behavior should be identical in this PR compared to master.
In HMD there shouldn't be any difference of interactivity or performances
we want to check in this PR is there is any latency issue that could have been introduced.
this means checking that tablet is rock solid in HMD, grabed objects move correctly with your hand and that there is no position tearing between different elements (entity vs overlay vs avatar joints) when they are attached together.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3782/HighFidelity-Beta-PR11499-0cbf19bcf141765c153befda56705359272b7c58.exe

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -188,7 +188,7 @@
 #include "InterfaceParentFinder.h"
 #include "ui/OctreeStatsProvider.h"
 
-#include "FrameTimingsScriptingInterface.h"
+//#include "FrameTimingsScriptingInterface.h"
```
Comment (samcake) [inline interface/src/Application.cpp]: This doesn't seem used at all, can we remove ?

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -2216,7 +2217,7 @@ void Application::initializeGL() {
     update(0);
 }
 
-FrameTimingsScriptingInterface _frameTimingsScriptingInterface;
+//FrameTimingsScriptingInterface _frameTimingsScriptingInterface;
```
Comment (samcake) [inline interface/src/Application.cpp]: This doesn't seem used at all, can we remove ?

File: interface/src/Application.cpp [hunk 3]
```diff
@@ -2272,7 +2273,7 @@ void Application::initializeUi() {
     surfaceContext->setContextProperty("Recording", DependencyManager::get<RecordingScriptingInterface>().data());
     surfaceContext->setContextProperty("Preferences", DependencyManager::get<Preferences>().data());
     surfaceContext->setContextProperty("AddressManager", DependencyManager::get<AddressManager>().data());
-    surfaceContext->setContextProperty("FrameTimings", &_frameTimingsScriptingInterface);
+    //surfaceContext->setContextProperty("FrameTimings", &_frameTimingsScriptingInterface); // TODO: Remove this Context Property ? i don;t see anywhere
```
Comment (samcake) [inline interface/src/Application.cpp]: This doesn't seem used at all, can we remove ?
this is the only place we do something with the _frameTimingsScriptiingInterface
and then it snot used anywhere ?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3785/HighFidelity-Beta-PR11499-6c510bc18f297842fac4c4c3e77dc20041a3b65c.dmg

File: interface/src/Application.cpp [hunk 4]
```diff
@@ -2377,301 +2378,105 @@ void Application::initializeUi() {
     offscreenSurfaceCache->reserve(Web3DOverlay::QML, 2);
 }
 
-void Application::paintGL() {
```
Comment (samcake) [inline interface/src/Application.cpp]: PaintGL moved to Application_render.cpp, but the work that is updating the camera or configuring the RenderArgs has been kept in the Game loop
either in "updateCamera"
or in the lambda called at the end of ::update()

File: interface/src/Application.cpp [hunk 5]
```diff
@@ -5230,9 +5036,123 @@ void Application::update(float deltaTime) {
         _postUpdateLambdas.clear();
     }
 
+    editRenderArgs([this](AppRenderArgs& appRenderArgs) {
```
Comment (samcake) [inline interface/src/Application.cpp]: This is where we do all the work that has been moved out from from paintGL or displaySide

File: interface/src/Application.cpp [hunk 6]
```diff
@@ -5230,9 +5036,123 @@ void Application::update(float deltaTime) {
...
+            appRenderArgs._renderArgs.setViewFrustum(_displayViewFrustum);
+        }
+    });
+
     AnimDebugDraw::getInstance().update();
 
     DependencyManager::get<LimitlessVoiceRecognitionScriptingInterface>()->update();
+
+    // Game loop is done, mark the end of the frame for the scene transactions and the render loop to take over
+    getMain3DScene()->enqueueFrame();
```
Comment (samcake) [inline interface/src/Application.cpp]: The transaction queue of the scene is now framed with this call in order to mark the end of the changes belonging to a specific game loop frame

File: interface/src/Application.cpp [hunk 7]
```diff
@@ -5522,116 +5441,6 @@ void Application::copyDisplayViewFrustum(ViewFrustum& viewOut) const {
     viewOut = _displayViewFrustum;
 }
 
-void Application::copyShadowViewFrustum(ViewFrustum& viewOut) const {
```
Comment (samcake) [inline interface/src/Application.cpp]: not used anymore

File: interface/src/Application.cpp [hunk 8]
```diff
@@ -5522,116 +5441,6 @@ void Application::copyDisplayViewFrustum(ViewFrustum& viewOut) const {
...
-    QMutexLocker viewLocker(&_viewMutex);
-    viewOut = _shadowViewFrustum;
-}
-
-// WorldBox Render Data & rendering functions
```
Comment (samcake) [inline interface/src/Application.cpp]: This WorldBOx has moved together with display side and paintGL to Application_render.cpp

File: interface/src/Application.h [hunk 9]
```diff
@@ -469,8 +470,6 @@ private slots:
 
     void queryOctree(NodeType_t serverType, PacketType packetType, NodeToJurisdictionMap& jurisdictions);
 
-    void renderRearViewMirror(RenderArgs* renderArgs, const QRect& region, bool isZoomed);
```
Comment (samcake) [inline interface/src/Application.h]: just the header was still there...

File: interface/src/Application.h [hunk 10]
```diff
@@ -480,7 +479,7 @@ private slots:
 
     void initializeAcceptedFiles();
 
-    void displaySide(RenderArgs* renderArgs, Camera& whichCamera, bool selfAvatarOnly = false);
```
Comment (samcake) [inline interface/src/Application.h]: i renamed displaySide to runRenderFrame...

File: interface/src/Application.h [hunk 11]
```diff
@@ -624,6 +622,24 @@ private slots:
     render::EnginePointer _renderEngine{ new render::Engine() };
     gpu::ContextPointer _gpuContext; // initialized during window creation
 
+    mutable QMutex _renderArgsMutex{ QMutex::Recursive };
+    struct AppRenderArgs {
```
Comment (samcake) [inline interface/src/Application.h]: A new struct containing all the values transferred from Game loop to Render loop

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3785/HighFidelity-Beta-PR11499-6c510bc18f297842fac4c4c3e77dc20041a3b65c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3787/HighFidelity-Beta-PR11499-ec10d38983f813e37ff123486602f0ea14fa2add.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3787/HighFidelity-Beta-PR11499-ec10d38983f813e37ff123486602f0ea14fa2add.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3830/HighFidelity-Beta-PR11499-bb99f68d40ec7e0a67bb366067380d2638377a59.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3830/HighFidelity-Beta-PR11499-bb99f68d40ec7e0a67bb366067380d2638377a59.exe

**Comment by YuppyPlays** [comment]
@samcake  if ready for QA could we get a test plan please :) or is this just smoke-test?

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3839/HighFidelity-Beta-PR11499-64f70f1ebbd3dcf69b772f5b2076466738ba2484.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3839/HighFidelity-Beta-PR11499-64f70f1ebbd3dcf69b772f5b2076466738ba2484.exe

**Comment by Nex-Pro** [comment]
T E S T I N G  this

**Comment by Nex-Pro** [comment]
### PASS
No issues seen with increased latency or shearing. Behaviour is same as latest dev build.

File: libraries/render/src/render/Scene.cpp [hunk 12]
```diff
@@ -101,16 +98,46 @@ void consolidateTransaction(TransactionQueue& queue, Transaction& singleBatch) {
         queue.pop();
     };
 }
- 
-void Scene::processTransactionQueue() {
+
+uint32_t Scene::enqueueFrame() {
```
Comment (AndrewMeadows) [inline libraries/render/src/render/Scene.cpp]: `Scene::enqueueFrame()` returns `uint32_t` but that return value is never used by any code that calls it.  Is this unused feature for debugging?

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3885/HighFidelity-Beta-PR11499-d7333d3fa34810311c2643d89527ddf8ad658197.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/3891/HighFidelity-Beta-PR11499-d773b40a63957d6701c9510a76cc4c84d14cc7dd.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/3891/HighFidelity-Beta-PR11499-d773b40a63957d6701c9510a76cc4c84d14cc7dd.exe

**Pull request merged**